csrc/fusion_profiler.h

```diff
@@ -0,0 +1,308 @@
+
+ public:
+  //! Method to access FusionProfiler singleton
+  static FusionProfiler* get();
+
+  void reset();
+  ProfilerState state() const;
+  SegmentProfiler& segment(size_t idx);
+
+  //! Profiling Methods
+  void start(bool cupti_disable = false);
+  void stop();
+  void createSegments(size_t num);
+  void startParallelCompile();
+  void stopParallelCompile();
+  void inputBytesAccessed(int64_t bytes);
+  void outputBytesAccessed(int64_t bytes);
+  const FusionProfile& profile() const;
+
+  //! Methods to capture Asynchronous CUPTI activity that get called from
+  //! functions registered with CUPTI.
+  //! Correlation ID -> Segment ID
+  void recordAsyncCorrIdActivity(uint32_t seg_id, uint32_t corr_id);
+  //! Collects CUPTI Kernel Activity
+  void recordAsyncKernelActivity(KernelProfile prof);
+  //! Ptr to the CUPTI Activity Buffer
+  uint8_t* cuptiBufferPtr();
+
+ public:
+  // CUPTI buffer size 4.0 KB
+  // The original example code used an 8MB buffer.  Such a larger buffer
+  // impacted host time overhead significantly.
+  static constexpr size_t cupti_activity_buffer_size{size_t(4 * 1024)};
+
+ private:
+  static FusionProfiler* singleton_;
+
+  //! Disables CUPTI usage in order to measure Host Time without CUPTI overhead
+  bool cupti_disabled_;
+  //! Buffer for Cupti to store Activity Buffers during async activity
+  std::vector<uint8_t> cupti_buffer_;
+  //! The state is used to check for errors in usage
+  ProfilerState state_;
+
+  //! Data members with information that is aggregated into a FusionProfile
+  int64_t fusion_id_;
+  bool parallel_compile_;
+  FusionProfile profile_;
+  CudaEventTimer fusion_timer_;
+  HostTimer host_timer_;
+  HostTimer parallel_compile_timer_;
+  std::vector<SegmentProfiler> segments_;
+  //! The FusionProfiler collects a cache of device descriptors so each segment
+  //! does not need to spend time re-generating the information.
+  std::vector<DeviceDescriptor> device_descriptors_;
+
+  //! These 2 data members are used to collect and connect asynchronous records,
+  //! generated by CUPTI, to the segments responsible for the activity
+  std::vector<KernelProfile> kernel_profiles_;
+  std::unordered_map<uint32_t, uint32_t> corrid_2_segid_;
+};
+
+//! The global functions below are distinct and could be separated out from the
+//! rest of the profiler code.  Their goal is to encapsulate boiler plate
+//! enablement code away from the code being profiled.
+
+//! A Fusion Profiler enabled check "decorator"
+template <class T>
+auto fusion_profiler_enabled(T&& func) {
+  // we create a closure below
+  auto new_function = [func = std::forward<T>(func)](auto&&... args) {
+    if (isDebugDumpEnabled(DebugDumpOption::FusionProfiler) ||
+        isDebugDumpEnabled(DebugDumpOption::FusionProfilerNocupti) ||
+        isDebugDumpEnabled(DebugDumpOption::FusionProfilerVerbose) ||
+        isOptionEnabled(EnableOption::FusionProfiler) ||
+        isOptionEnabled(EnableOption::FusionProfilerNocupti)) {
```

Comment:
An easier and cleaner way is to let `FusionProfiler::get()` return `nullptr` when FusionProfiler is not enabled and then check just for that. This way, we only need to decide once whether FusionProfiler is enabled from EnableOptions.

Reply:
I removed the prevous form of enablement. The new enablement is via:
```
isProfilerEnabled()
```